src/Plutarch/Test/Export.hs

```diff
@@ -0,0 +1,726 @@
+import Optics.Optic ((%), (%%))
+import Optics.Setter (over, set)
+import Optics.TH (makeFieldLabelsNoPrefix)
+import Plutarch.Evaluate (evalScriptHuge)
+import PlutusLedgerApi.V1.Scripts (applyArguments)
+import PlutusLedgerApi.V2 (
+  Data,
+  Script,
+  ScriptContext,
+  ToData,
+  toData,
+ )
+import Ply (
+  ScriptRole (MintingPolicyRole, ValidatorRole),
+  TypedScriptEnvelope (tsRole, tsScript),
+ )
+import ScriptExport.ScriptInfo (
+  Linker,
+  LinkerError,
+  RawScriptExport,
+  ScriptExport,
+  runLinker,
+ )
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.Providers (
+  IsTest (run, testOptions),
+  singleTest,
+  testFailed,
+  testPassed,
+ )
+
+{- | Typed parameters that are given to a script when testing. For minting
+ policies, the redeemer is ignored; you can set it to '()' to keep the
+ compiler happy.
+
+ @since 1.2.1
+-}
+data ScriptParams (dat :: Type) (red :: Type) = ScriptParams
+  { datum :: dat
+  , redeemer :: red
+  , context :: ScriptContext
+  }
+
+makeFieldLabelsNoPrefix ''ScriptParams
+
+{- | A combination of (initial) script parameters, as well as any required
+ parameters that should be passed to the linker when assembling tests.
+
+ @since 1.2.1
+-}
+data TestParams (dat :: Type) (red :: Type) (p :: Type) = TestParams
+  { scriptParams :: ScriptParams dat red
+  , linkerParams :: p
+  }
+
+makeFieldLabelsNoPrefix ''TestParams
+
+{- | What you expect to happen when a script is tested with specified
+ parameters.
+
+ @since 1.2.1
+-}
+data ScriptOutcome = Crashes | Runs
+  deriving stock
+    ( -- | @since 1.2.1
+      Eq
+    , -- | @since 1.2.1
+      Ord
+    , -- | @since 1.2.1
+      Show
+    )
+
+{- | A helper newtype for passing script arguments implicitly, as well as
+ modifying them using local scoping.
+
+ The intended use is with @do@-notation, using a combination of contextual
+ argument modifiers and assertions, then \'running\' with 'exportTests':
+
+ > tests :: TestTree
+ > tests = exportTests "my tests" export linker testParams $ do
+ >    assertValidator Crashes "my context is bad" "script1"
+ >    modifyingSC fixUpContext $ do
+ >      assertValidator Crashes "context fine, redeemer bad" "script1"
+ >      modifyingRedeemer fixUpRedeemer $ do
+ >        assertValidator Runs "now everything is fine" "script1"
+ >      -- Redeemer fix-up resets here
+ >      assertMintingPolicy Crashes "inappropriate datum" "script2"
+ >      settingDatum' redeemerDatum $ do
+ >        assertMintingPolicy Runs "appropriate datum" "script2"
+
+ You can also define separate \'bits\' of a suite using 'WithExport' directly.
+
+ @since 1.2.1
+-}
+newtype WithExport (dat :: Type) (red :: Type) (info :: Type) (a :: Type)
+  = WithExport (RWS (WithExportEnv dat red info) WithExportLog () a)
+  deriving
+    ( -- | @since 1.2.1
+      Functor
+    , -- | @since 1.2.1
+      Applicative
+    , -- | @since 1.2.1
+      Monad
+    )
+    via ( RWS (WithExportEnv dat red info) WithExportLog ()
+        )
```

Comment:
This is a good point, we can make `dat` and `red` just a phantom type and only carry `Data` underneath. That way we can make interface that bypasses checking type for datum and redeemer

Reply:
That makes local modifications _much_ more annoying, since I'd have to convert to and from all the time. I do agree with @t1lde though, it is quite restrictive.